applications/FluidDynamicsApplication/custom_elements/vms_adjoint_element.h

```diff
@@ -569,6 +751,146 @@ class KRATOS_API(FLUID_DYNAMICS_APPLICATION) VMSAdjointElement: public Element
+        unsigned int FirstRow(0), FirstCol(0);
+        double K; // Temporary results
+
+        // If we want to use more than one Gauss point to integrate the convective term, this has to be evaluated once per integration point
+        const array_1d<double, TNumNodes> AGradN = prod(rShapeDeriv, rAdvVel);
+
+        // Note: Dof order is (vx,vy,[vz,]p) for each node
+        for (unsigned int i = 0; i < TNumNodes; ++i)
+        {
+            // Loop over columns
+            for (unsigned int j = 0; j < TNumNodes; ++j)
+            {
+                // Delta(u) * TauOne * [ AdvVel * Grad(v) ] in velocity block
+                K = Coef * Density * AGradN[i] * Density * rShapeFunc[j];
+
+                for (unsigned int d = 0; d < TDim; ++d) // iterate over dimensions for velocity Dofs in this node combination
+                {
+                    rLHSMatrix(FirstRow + d, FirstCol + d) += K;
+                    // Delta(u) * TauOne * Grad(q) in q * Div(u) block
+                    rLHSMatrix(FirstRow + TDim, FirstCol + d) += Coef * Density * rShapeDeriv(i, d) * rShapeFunc[j];
+                }
+                // Update column index
+                FirstCol += BlockSize;
+            }
+            // Update matrix indices
+            FirstRow += BlockSize;
+            FirstCol = 0;
+        }
+    }
+
+    void AddIntegrationPointVelocityContribution(
+        MatrixType& rDampingMatrix,
+        VectorType& rDampRHS,
+        const double Density,
+        const double Viscosity,
+        const array_1d<double, TDim>& rAdvVel,
+        const array_1d<double, TDim>& rBodyForce,
+        const double TauOne,
+        const double TauTwo,
+        const array_1d<double, TNumNodes>& rShapeFunc,
+        const BoundedMatrix<double, TNumNodes, TDim>& rShapeDeriv,
+        const double Weight)
+    {
+        // If we want to use more than one Gauss point to integrate the convective term, this has to be evaluated once per integration point
+        const array_1d<double, TNumNodes> AGradN = prod(rShapeDeriv, rAdvVel);
+
+        // Build the local matrix and RHS
+        unsigned int FirstRow(0),
+            FirstCol(0); // position of the first term of the local matrix that corresponds to each node combination
+        double K, G, PDivV, L, qF; // Temporary results
+
+        array_1d<double, TDim> BodyForce = rBodyForce * Density;
+
+        for (unsigned int i = 0; i < TNumNodes; ++i) // iterate over rows
+        {
+            for (unsigned int j = 0; j < TNumNodes; ++j) // iterate over columns
+            {
+                // Calculate the part of the contributions that is constant for each node combination
+
+                // Velocity block
+                K = Density * rShapeFunc[i] * AGradN[j]; // Convective term: v * ( a * Grad(u) )
+                // K = 0.5 * Density * (rShapeFunc[i] * AGradN[j] - AGradN[i] * rShapeFunc[j]); // Skew-symmetric convective term 1/2( v*grad(u)*u - grad(v) uu )
```

Comment:
As we are not using the skew-symmetric form, I'd remove this to avoid misunderstandings.
```suggestion
```

Reply:
done :)